tools/system_libs.py

```diff
@@ -69,7 +69,18 @@ def calculate(temp_files, in_temp, stdout_, stderr_, forced=[]):
 
   def read_symbols(path):
     with open(path) as f:
-      return shared.Building.parse_symbols(f.read()).defs
+      content = f.read()
+
+      # Require that Windows newlines should not be present in a symbols file, if running on Linux or macOS
+      # This kind of mismatch can occur if one copies a zip file of Emscripten cloned on Windows over to
+      # a Linux or macOS system. It will result in Emscripten linker getting confused on stray \r characters,
+      # and be unable to link any library symbols properly. We could harden against this by .strip()ping the
+      # opened files, but it is possible that the mismatching line endings can cause random problems elsewhere
+      # in the toolchain, hence abort execution if so.
+      if os.name != 'nt' and '\r\n' in content:
+        raise Exception('Windows newlines \\r\\n detected in symbols file "' + path + '"! This could happen for example when copying Emscripten checkout from Windows to Linux or macOS. Please use Unix line endings on checkouts of Emscripten on Linux and macOS!')
+
```

Comment:
how about adding a method in `shared.py` that receives some text and does this text, and returns it, so that here we could write `return shared.Building.parse_symbols(verify_newlines(f.read())).defs`?

Reply:
Hmm, I'm not sure if that would gain much - it would just spread out the code across multiple files to make it perhaps harder to follow?

Note that the intent here is not that we would diligently start checking all places for consistent newlines, but just here, since this place is the one where the result is catastrophic when such an error occurs, so good to have a check close to that, and it is likely that if one text file in the repository has wrong newlines, then probably all of them will have, so checking this place should be enough to detect if a repository has been accidentally been obtained with mixed newlines.

Reply:
In that case, should the check not be in `parse_symbols`, so that it covers other callers to the same place?

lgtm with it there (or maybe I missed the point here?)

Reply:
I could move it inside `parse_symbols`, but I don't think that would be needed here. The only other use of `parse_symbols` is in

```
  @staticmethod
  def llvm_nm_uncached(filename, stdout=PIPE, stderr=PIPE, include_internal=False):
    # LLVM binary ==> list of symbols
    proc = run_process([LLVM_NM, filename], stdout=stdout, stderr=stderr, check=False)
    if proc.returncode == 0:
      return Building.parse_symbols(proc.stdout, include_internal)
```
but that code calls `llvm-nm` locally, it is not possible that call would ever produce incorrect newlines, since it is result of local execution. Extending the check to that would just have the effect of slowing down compilation? I can do it either way, I think the current check will be enough to detect bad zips copied from Windows to Linux/macOS, with minimal influence overall?

Reply:
Fair enough, lgtm.